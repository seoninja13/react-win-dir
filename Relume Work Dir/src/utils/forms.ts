/**
 * Validate an email address
 */
export function validateEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

/**
 * Validate a phone number
 */
export function validatePhone(phone: string): boolean {
  const phoneRegex = /^\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$/;
  return phoneRegex.test(phone);
}

/**
 * Validate a ZIP code
 */
export function validateZip(zip: string): boolean {
  const zipRegex = /^\d{5}(-\d{4})?$/;
  return zipRegex.test(zip);
}

/**
 * Validate a name (first or last)
 */
export function validateName(name: string): boolean {
  return name.trim().length >= 2;
}

/**
 * Validate a message
 */
export function validateMessage(message: string): boolean {
  return message.trim().length >= 10;
}

/**
 * Validate a lead form
 */
export function validateLeadForm(formData: {
  first_name: string;
  last_name: string;
  email: string;
  phone?: string;
  zip?: string;
  message?: string;
}): { isValid: boolean; errors: Record<string, string> } {
  const errors: Record<string, string> = {};
  
  if (!validateName(formData.first_name)) {
    errors.first_name = 'First name is required and must be at least 2 characters';
  }
  
  if (!validateName(formData.last_name)) {
    errors.last_name = 'Last name is required and must be at least 2 characters';
  }
  
  if (!validateEmail(formData.email)) {
    errors.email = 'Please enter a valid email address';
  }
  
  if (formData.phone && !validatePhone(formData.phone)) {
    errors.phone = 'Please enter a valid phone number';
  }
  
  if (formData.zip && !validateZip(formData.zip)) {
    errors.zip = 'Please enter a valid ZIP code';
  }
  
  if (formData.message && !validateMessage(formData.message)) {
    errors.message = 'Message must be at least 10 characters';
  }
  
  return {
    isValid: Object.keys(errors).length === 0,
    errors,
  };
}

/**
 * Submit a lead form to the API
 */
export async function submitLeadForm(formData: any) {
  const response = await fetch('/api/leads', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(formData),
  });
  
  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.error || 'Failed to submit form');
  }
  
  return response.json();
}
