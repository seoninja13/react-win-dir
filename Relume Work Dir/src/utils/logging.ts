import { supabase } from './supabase';

type LogLevel = 'info' | 'warning' | 'error' | 'debug';

export async function log(
  level: LogLevel,
  message: string,
  details?: any,
  tags?: string[]
) {
  try {
    const { error } = await supabase.from('logs').insert([
      {
        level,
        message,
        details,
        tags,
        source: 'client',
        url: typeof window !== 'undefined' ? window.location.href : '',
        user_agent: typeof navigator !== 'undefined' ? navigator.userAgent : '',
      },
    ]);
    
    if (error) {
      console.error('Failed to log message:', error);
    }
    
    // Also log to console in development
    if (process.env.NODE_ENV === 'development') {
      console[level === 'error' ? 'error' : level === 'warning' ? 'warn' : 'log'](
        `[${level.toUpperCase()}] ${message}`,
        details
      );
    }
  } catch (err) {
    console.error('Logging failed:', err);
  }
}

export function logInfo(message: string, details?: any, tags?: string[]) {
  return log('info', message, details, tags);
}

export function logWarning(message: string, details?: any, tags?: string[]) {
  return log('warning', message, details, tags);
}

export function logError(message: string, details?: any, tags?: string[]) {
  // Also log to recent_errors table
  try {
    supabase.from('recent_errors').insert([
      {
        message,
        details,
        source: 'client',
        url: typeof window !== 'undefined' ? window.location.href : '',
      },
    ]);
  } catch (err) {
    console.error('Error logging to recent_errors failed:', err);
  }
  
  return log('error', message, details, tags);
}

export function logDebug(message: string, details?: any, tags?: string[]) {
  return log('debug', message, details, tags);
}
