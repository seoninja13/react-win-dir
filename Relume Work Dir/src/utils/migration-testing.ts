/**
 * Migration Testing Utility
 * 
 * This utility provides functions for testing pages during the App Router migration.
 * It includes functions for testing routing, functionality, and performance.
 */

import { logger } from './logger';

interface RouteTestResult {
  route: string;
  status: 'success' | 'failure';
  statusCode?: number;
  error?: string;
}

interface FunctionalityTestResult {
  test: string;
  status: 'success' | 'failure';
  error?: string;
}

interface PerformanceTestResult {
  metric: string;
  value: number;
  unit: string;
  threshold?: number;
  status: 'success' | 'failure';
}

/**
 * Tests a route by making a fetch request to it
 * @param route The route to test
 * @returns A promise that resolves to a RouteTestResult
 */
export async function testRoute(route: string): Promise<RouteTestResult> {
  try {
    const response = await fetch(route);
    const status = response.ok ? 'success' : 'failure';
    
    logger.logMigrationTest(`Route test: ${route}`, status === 'success' ? 'passed' : 'failed', {
      statusCode: response.status,
      route,
    });
    
    return {
      route,
      status,
      statusCode: response.status,
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    
    logger.logMigrationTest(`Route test: ${route}`, 'failed', {
      error: errorMessage,
      route,
    });
    
    return {
      route,
      status: 'failure',
      error: errorMessage,
    };
  }
}

/**
 * Tests multiple routes
 * @param routes An array of routes to test
 * @returns A promise that resolves to an array of RouteTestResult
 */
export async function testRoutes(routes: string[]): Promise<RouteTestResult[]> {
  const results: RouteTestResult[] = [];
  
  for (const route of routes) {
    const result = await testRoute(route);
    results.push(result);
  }
  
  return results;
}

/**
 * Tests a specific functionality
 * @param testName The name of the test
 * @param testFn The test function that returns a boolean or throws an error
 * @returns A promise that resolves to a FunctionalityTestResult
 */
export async function testFunctionality(
  testName: string,
  testFn: () => boolean | Promise<boolean>
): Promise<FunctionalityTestResult> {
  try {
    const result = await testFn();
    const status = result ? 'success' : 'failure';
    
    logger.logMigrationTest(`Functionality test: ${testName}`, status === 'success' ? 'passed' : 'failed');
    
    return {
      test: testName,
      status,
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    
    logger.logMigrationTest(`Functionality test: ${testName}`, 'failed', {
      error: errorMessage,
    });
    
    return {
      test: testName,
      status: 'failure',
      error: errorMessage,
    };
  }
}

/**
 * Measures the performance of a function
 * @param metricName The name of the metric
 * @param fn The function to measure
 * @param unit The unit of measurement
 * @param threshold Optional threshold for success/failure
 * @returns A promise that resolves to a PerformanceTestResult
 */
export async function measurePerformance(
  metricName: string,
  fn: () => any | Promise<any>,
  unit: string,
  threshold?: number
): Promise<PerformanceTestResult> {
  try {
    const start = performance.now();
    await fn();
    const end = performance.now();
    const value = end - start;
    
    const status = threshold ? (value <= threshold ? 'success' : 'failure') : 'success';
    
    logger.logMigrationTest(`Performance test: ${metricName}`, status === 'success' ? 'passed' : 'failed', {
      value,
      unit,
      threshold,
    });
    
    return {
      metric: metricName,
      value,
      unit,
      threshold,
      status,
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    
    logger.logMigrationTest(`Performance test: ${metricName}`, 'failed', {
      error: errorMessage,
    });
    
    return {
      metric: metricName,
      value: 0,
      unit,
      threshold,
      status: 'failure',
    };
  }
}

/**
 * Runs a complete test suite for a page
 * @param pageName The name of the page
 * @param route The route to the page
 * @param functionalityTests An array of functionality tests
 * @returns A promise that resolves to an object with test results
 */
export async function testPage(
  pageName: string,
  route: string,
  functionalityTests: Array<{
    name: string;
    test: () => boolean | Promise<boolean>;
  }> = []
) {
  logger.migration(`Starting tests for page: ${pageName}`, {
    pageName,
    route,
    timestamp: new Date().toISOString(),
  });
  
  // Test route
  const routeResult = await testRoute(route);
  
  // Test functionality
  const functionalityResults: FunctionalityTestResult[] = [];
  for (const test of functionalityTests) {
    const result = await testFunctionality(test.name, test.test);
    functionalityResults.push(result);
  }
  
  // Measure performance
  const performanceResult = await measurePerformance(
    'Page Load Time',
    async () => {
      await fetch(route);
    },
    'ms',
    1000 // 1 second threshold
  );
  
  // Log summary
  const routeSuccess = routeResult.status === 'success';
  const functionalitySuccess = functionalityResults.every(result => result.status === 'success');
  const performanceSuccess = performanceResult.status === 'success';
  const overallSuccess = routeSuccess && functionalitySuccess && performanceSuccess;
  
  logger.migration(`Test summary for page: ${pageName}`, {
    pageName,
    route,
    overallStatus: overallSuccess ? 'passed' : 'failed',
    routeStatus: routeSuccess ? 'passed' : 'failed',
    functionalityStatus: functionalitySuccess ? 'passed' : 'failed',
    performanceStatus: performanceSuccess ? 'passed' : 'failed',
    timestamp: new Date().toISOString(),
  });
  
  return {
    pageName,
    route,
    overallSuccess,
    routeResult,
    functionalityResults,
    performanceResult,
  };
}

/**
 * Creates a test page that can be used to test the migration
 * @param pageName The name of the page
 * @param tests An object with test functions
 * @returns A React component that can be used to test the migration
 */
export function createTestPage(pageName: string, tests: Record<string, () => boolean | Promise<boolean>>) {
  return function TestPage() {
    const runTests = async () => {
      const results: Record<string, boolean> = {};
      
      for (const [testName, testFn] of Object.entries(tests)) {
        try {
          const result = await testFn();
          results[testName] = result;
          logger.logMigrationTest(`${testName}`, result ? 'passed' : 'failed', {
            pageName,
          });
        } catch (error) {
          results[testName] = false;
          logger.logMigrationTest(`${testName}`, 'failed', {
            pageName,
            error: error instanceof Error ? error.message : String(error),
          });
        }
      }
      
      return results;
    };
    
    return {
      pageName,
      runTests,
    };
  };
}
