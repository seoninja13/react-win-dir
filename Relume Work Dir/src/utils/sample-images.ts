/**
 * Sample Images Utility
 * 
 * This file contains utilities for copying sample images to the public directory
 * and generating image URLs for use in components.
 */

import fs from 'fs';
import path from 'path';

// Define source and destination paths
const SAMPLE_IMAGES_DIR = path.join(process.cwd(), '..', 'Sample-images');
const PUBLIC_IMAGES_DIR = path.join(process.cwd(), 'public', 'images');

// Define image categories
const IMAGE_CATEGORIES = ['Windows', 'Doors'];

/**
 * Copy sample images to the public directory
 * 
 * @returns An object with arrays of image paths for each category
 */
export async function copySampleImages(): Promise<Record<string, string[]>> {
  const result: Record<string, string[]> = {};

  try {
    // Create public/images directory if it doesn't exist
    if (!fs.existsSync(PUBLIC_IMAGES_DIR)) {
      fs.mkdirSync(PUBLIC_IMAGES_DIR, { recursive: true });
    }

    // Process each category
    for (const category of IMAGE_CATEGORIES) {
      const sourceCategoryDir = path.join(SAMPLE_IMAGES_DIR, category);
      const destCategoryDir = path.join(PUBLIC_IMAGES_DIR, category.toLowerCase());

      // Skip if source directory doesn't exist
      if (!fs.existsSync(sourceCategoryDir)) {
        console.warn(`Source directory not found: ${sourceCategoryDir}`);
        continue;
      }

      // Create destination category directory if it doesn't exist
      if (!fs.existsSync(destCategoryDir)) {
        fs.mkdirSync(destCategoryDir, { recursive: true });
      }

      // Get list of image files
      const files = fs.readdirSync(sourceCategoryDir);
      const imagePaths: string[] = [];

      // Copy each file
      for (const file of files) {
        const sourceFilePath = path.join(sourceCategoryDir, file);
        const destFilePath = path.join(destCategoryDir, file);

        // Skip if not a file
        if (!fs.statSync(sourceFilePath).isFile()) {
          continue;
        }

        // Copy file if it doesn't exist or is newer
        if (
          !fs.existsSync(destFilePath) ||
          fs.statSync(sourceFilePath).mtime > fs.statSync(destFilePath).mtime
        ) {
          fs.copyFileSync(sourceFilePath, destFilePath);
          console.log(`Copied ${sourceFilePath} to ${destFilePath}`);
        }

        // Add to result
        imagePaths.push(`/images/${category.toLowerCase()}/${file}`);
      }

      result[category.toLowerCase()] = imagePaths;
    }

    return result;
  } catch (error) {
    console.error('Error copying sample images:', error);
    throw error;
  }
}

/**
 * Get sample image URLs for a specific category
 * 
 * @param category - The category to get images for (e.g., 'windows', 'doors')
 * @returns An array of image URLs
 */
export function getSampleImageUrls(category: string): string[] {
  try {
    const categoryDir = path.join(PUBLIC_IMAGES_DIR, category.toLowerCase());

    // Return empty array if directory doesn't exist
    if (!fs.existsSync(categoryDir)) {
      return [];
    }

    // Get list of image files
    const files = fs.readdirSync(categoryDir);

    // Filter for image files and create URLs
    return files
      .filter(file => /\.(jpg|jpeg|png|gif|webp)$/i.test(file))
      .map(file => `/images/${category.toLowerCase()}/${file}`);
  } catch (error) {
    console.error(`Error getting sample image URLs for ${category}:`, error);
    return [];
  }
}

/**
 * Get a random sample image URL for a specific category
 * 
 * @param category - The category to get an image for (e.g., 'windows', 'doors')
 * @returns A random image URL or null if no images are available
 */
export function getRandomSampleImageUrl(category: string): string | null {
  const imageUrls = getSampleImageUrls(category);

  if (imageUrls.length === 0) {
    return null;
  }

  const randomIndex = Math.floor(Math.random() * imageUrls.length);
  return imageUrls[randomIndex];
}

/**
 * Get a specific number of sample image URLs for a category
 * 
 * @param category - The category to get images for (e.g., 'windows', 'doors')
 * @param count - The number of images to get
 * @returns An array of image URLs
 */
export function getSampleImageUrlsCount(category: string, count: number): string[] {
  const imageUrls = getSampleImageUrls(category);

  if (imageUrls.length <= count) {
    return imageUrls;
  }

  // Shuffle array and take first 'count' elements
  return [...imageUrls]
    .sort(() => 0.5 - Math.random())
    .slice(0, count);
}
