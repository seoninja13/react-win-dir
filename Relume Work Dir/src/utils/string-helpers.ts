/**
 * String Helper Utilities
 * 
 * A collection of utility functions for string manipulation and validation.
 * These functions are commonly used throughout the application.
 */

/**
 * Capitalizes the first letter of a string
 * @param str - The string to capitalize
 * @returns The string with the first letter capitalized
 */
export function capitalize(str: string): string {
  if (!str || typeof str !== 'string') {
    return '';
  }
  return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

/**
 * Converts a string to kebab-case
 * @param str - The string to convert
 * @returns The string in kebab-case format
 */
export function toKebabCase(str: string): string {
  if (!str || typeof str !== 'string') {
    return '';
  }
  return str
    .replace(/([a-z])([A-Z])/g, '$1-$2')
    .replace(/[\s_]+/g, '-')
    .toLowerCase();
}

/**
 * Converts a string to camelCase
 * @param str - The string to convert
 * @returns The string in camelCase format
 */
export function toCamelCase(str: string): string {
  if (!str || typeof str !== 'string') {
    return '';
  }
  return str
    .replace(/[-_\s]+(.)?/g, (_, char) => char ? char.toUpperCase() : '')
    .replace(/^[A-Z]/, char => char.toLowerCase());
}

/**
 * Truncates a string to a specified length and adds ellipsis
 * @param str - The string to truncate
 * @param maxLength - Maximum length before truncation
 * @param suffix - Suffix to add when truncated (default: '...')
 * @returns The truncated string
 */
export function truncate(str: string, maxLength: number, suffix: string = '...'): string {
  if (!str || typeof str !== 'string') {
    return '';
  }
  if (str.length <= maxLength) {
    return str;
  }
  return str.slice(0, maxLength - suffix.length) + suffix;
}

/**
 * Removes HTML tags from a string
 * @param str - The string containing HTML
 * @returns The string with HTML tags removed
 */
export function stripHtml(str: string): string {
  if (!str || typeof str !== 'string') {
    return '';
  }
  return str.replace(/<[^>]*>/g, '');
}

/**
 * Validates if a string is a valid email address
 * @param email - The email string to validate
 * @returns True if valid email, false otherwise
 */
export function isValidEmail(email: string): boolean {
  if (!email || typeof email !== 'string') {
    return false;
  }
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

/**
 * Generates a random string of specified length
 * @param length - Length of the random string
 * @param charset - Character set to use (default: alphanumeric)
 * @returns Random string
 */
export function generateRandomString(
  length: number, 
  charset: string = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
): string {
  if (length <= 0) {
    return '';
  }
  let result = '';
  for (let i = 0; i < length; i++) {
    result += charset.charAt(Math.floor(Math.random() * charset.length));
  }
  return result;
}

/**
 * Counts the number of words in a string
 * @param str - The string to count words in
 * @returns Number of words
 */
export function wordCount(str: string): number {
  if (!str || typeof str !== 'string') {
    return 0;
  }
  return str.trim().split(/\s+/).filter(word => word.length > 0).length;
}

/**
 * Extracts URLs from a string
 * @param str - The string to extract URLs from
 * @returns Array of URLs found in the string
 */
export function extractUrls(str: string): string[] {
  if (!str || typeof str !== 'string') {
    return [];
  }
  const urlRegex = /https?:\/\/[^\s]+/g;
  return str.match(urlRegex) || [];
}

/**
 * Formats a string as a phone number
 * @param str - The string containing digits
 * @returns Formatted phone number or empty string if invalid
 */
export function formatPhoneNumber(str: string): string {
  if (!str || typeof str !== 'string') {
    return '';
  }
  const digits = str.replace(/\D/g, '');
  if (digits.length === 10) {
    return `(${digits.slice(0, 3)}) ${digits.slice(3, 6)}-${digits.slice(6)}`;
  }
  if (digits.length === 11 && digits[0] === '1') {
    return `+1 (${digits.slice(1, 4)}) ${digits.slice(4, 7)}-${digits.slice(7)}`;
  }
  return str; // Return original if can't format
}
