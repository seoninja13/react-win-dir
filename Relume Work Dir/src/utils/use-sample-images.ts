/**
 * Sample Images Hook
 * 
 * This hook provides access to sample images for use in components.
 * It simulates loading images from Supabase storage.
 */

import { useState, useEffect } from 'react';

// Define image categories
export type ImageCategory = 'windows' | 'doors' | 'siding' | 'roofing';

// Define image object
export interface SampleImage {
  id: string;
  name: string;
  url: string;
  category: ImageCategory;
  alt: string;
}

// Sample image data
const SAMPLE_IMAGES: Record<ImageCategory, SampleImage[]> = {
  windows: Array.from({ length: 10 }, (_, i) => ({
    id: `window-${i + 1}`,
    name: `windows${i + 1}.jpg`,
    url: `https://placehold.co/800x600?text=Window+Image+${i + 1}`,
    category: 'windows',
    alt: `Window sample image ${i + 1}`
  })),
  doors: Array.from({ length: 10 }, (_, i) => ({
    id: `door-${i + 1}`,
    name: `luxury-house-doors(${i + 1}).jpg`,
    url: `https://placehold.co/800x600?text=Door+Image+${i + 1}`,
    category: 'doors',
    alt: `Door sample image ${i + 1}`
  })),
  siding: Array.from({ length: 5 }, (_, i) => ({
    id: `siding-${i + 1}`,
    name: `siding${i + 1}.jpg`,
    url: `https://placehold.co/800x600?text=Siding+Image+${i + 1}`,
    category: 'siding',
    alt: `Siding sample image ${i + 1}`
  })),
  roofing: Array.from({ length: 5 }, (_, i) => ({
    id: `roofing-${i + 1}`,
    name: `roofing${i + 1}.jpg`,
    url: `https://placehold.co/800x600?text=Roofing+Image+${i + 1}`,
    category: 'roofing',
    alt: `Roofing sample image ${i + 1}`
  }))
};

/**
 * Hook to get sample images for a specific category
 * 
 * @param category - The category to get images for
 * @param count - The number of images to get (default: all)
 * @returns An array of sample images and loading state
 */
export function useSampleImages(category: ImageCategory, count?: number) {
  const [images, setImages] = useState<SampleImage[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const loadImages = async () => {
      try {
        setLoading(true);
        setError(null);

        // Simulate API call
        await new Promise(resolve => setTimeout(resolve, 500));

        // Get images for the specified category
        const categoryImages = SAMPLE_IMAGES[category] || [];
        
        // Limit to count if specified
        const limitedImages = count ? categoryImages.slice(0, count) : categoryImages;
        
        setImages(limitedImages);
      } catch (err) {
        setError(err instanceof Error ? err.message : 'An error occurred while loading images');
      } finally {
        setLoading(false);
      }
    };

    loadImages();
  }, [category, count]);

  return { images, loading, error };
}

/**
 * Get a random sample image for a specific category
 * 
 * @param category - The category to get an image for
 * @returns A random sample image
 */
export function getRandomSampleImage(category: ImageCategory): SampleImage {
  const categoryImages = SAMPLE_IMAGES[category] || [];
  const randomIndex = Math.floor(Math.random() * categoryImages.length);
  return categoryImages[randomIndex];
}

/**
 * Get a specific number of sample images for a category
 * 
 * @param category - The category to get images for
 * @param count - The number of images to get
 * @returns An array of sample images
 */
export function getSampleImages(category: ImageCategory, count: number): SampleImage[] {
  const categoryImages = SAMPLE_IMAGES[category] || [];
  
  if (categoryImages.length <= count) {
    return categoryImages;
  }
  
  // Shuffle array and take first 'count' elements
  return [...categoryImages]
    .sort(() => 0.5 - Math.random())
    .slice(0, count);
}
