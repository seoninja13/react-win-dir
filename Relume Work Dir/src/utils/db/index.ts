import { supabase } from '../supabase';
import { logError } from '../logging';

export async function handleDatabaseError(error: any, operation: string, details?: any) {
  const errorDetails = {
    operation,
    details,
    error: error.message,
    code: error.code,
  };
  
  console.error('Database operation failed:', errorDetails);
  
  // Log to Supabase if logging utility is available
  try {
    logError('Database operation failed', errorDetails);
  } catch (logError) {
    // If logging fails, just continue
    console.error('Failed to log error:', logError);
  }
  
  throw new Error(`Database operation failed: ${operation}`);
}

export { supabase };
