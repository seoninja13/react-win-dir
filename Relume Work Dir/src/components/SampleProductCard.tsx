'use client';

import { getRandomSampleImage, ImageCategory } from '@/utils/use-sample-images';
import Link from 'next/link';

interface SampleProductCardProps {
  title: string;
  description: string;
  category: ImageCategory;
  href: string;
  className?: string;
}

/**
 * Sample Product Card Component
 * 
 * This component displays a product card with an image, title, and description.
 * It uses a random sample image from the specified category.
 */
export default function SampleProductCard({ 
  title, 
  description, 
  category, 
  href,
  className = '' 
}: SampleProductCardProps) {
  const image = getRandomSampleImage(category);

  return (
    <div className={`bg-white rounded-lg overflow-hidden shadow-md hover:shadow-lg transition-shadow ${className}`}>
      {/* Image */}
      <div className="aspect-video overflow-hidden">
        <img
          src={image.url}
          alt={`${title} - ${image.alt}`}
          className="w-full h-full object-cover transition-transform hover:scale-105"
        />
      </div>
      
      {/* Content */}
      <div className="p-4">
        <h3 className="text-xl font-semibold mb-2">{title}</h3>
        <p className="text-gray-600 mb-4 line-clamp-3">{description}</p>
        <Link 
          href={href}
          className="inline-block px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition-colors"
        >
          Learn More
        </Link>
      </div>
    </div>
  );
}
