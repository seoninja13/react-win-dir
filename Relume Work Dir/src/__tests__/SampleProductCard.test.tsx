import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import SampleProductCard from '../components/SampleProductCard';
import { ImageCategory } from '@/utils/use-sample-images';

// Mock the use-sample-images utility
jest.mock('@/utils/use-sample-images', () => ({
  getRandomSampleImage: jest.fn(() => ({
    url: 'https://example.com/test-image.jpg',
    alt: 'Test image description'
  })),
  ImageCategory: {
    WINDOWS: 'windows',
    DOORS: 'doors',
    SIDING: 'siding',
    ROOFING: 'roofing'
  }
}));

// Mock Next.js Link component
jest.mock('next/link', () => {
  return function MockLink({ children, href, ...props }: any) {
    return <a href={href} {...props}>{children}</a>;
  };
});

describe('SampleProductCard', () => {
  const defaultProps = {
    title: 'Test Product',
    description: 'This is a test product description that should be displayed in the card.',
    category: 'windows' as ImageCategory,
    href: '/test-product'
  };

  // Basic rendering test
  it('renders without crashing', () => {
    render(<SampleProductCard {...defaultProps} />);
  });

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(<SampleProductCard {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  // Props rendering test
  it('renders with correct title and description', () => {
    render(<SampleProductCard {...defaultProps} />);
    
    expect(screen.getByText('Test Product')).toBeInTheDocument();
    expect(screen.getByText('This is a test product description that should be displayed in the card.')).toBeInTheDocument();
  });

  // Image rendering test
  it('renders image with correct attributes', () => {
    render(<SampleProductCard {...defaultProps} />);
    
    const image = screen.getByRole('img');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', 'https://example.com/test-image.jpg');
    expect(image).toHaveAttribute('alt', 'Test Product - Test image description');
  });

  // Link rendering test
  it('renders link with correct href', () => {
    render(<SampleProductCard {...defaultProps} />);
    
    const link = screen.getByRole('link', { name: 'Learn More' });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/test-product');
  });

  // Custom className test
  it('applies custom className when provided', () => {
    const customClass = 'custom-test-class';
    const { container } = render(
      <SampleProductCard {...defaultProps} className={customClass} />
    );
    
    const cardElement = container.firstChild as HTMLElement;
    expect(cardElement).toHaveClass(customClass);
  });

  // Default className test
  it('applies default classes when no custom className provided', () => {
    const { container } = render(<SampleProductCard {...defaultProps} />);
    
    const cardElement = container.firstChild as HTMLElement;
    expect(cardElement).toHaveClass('bg-white', 'rounded-lg', 'overflow-hidden', 'shadow-md');
  });

  // Hover interaction test
  it('has hover effects on image and link', () => {
    render(<SampleProductCard {...defaultProps} />);
    
    const image = screen.getByRole('img');
    const link = screen.getByRole('link', { name: 'Learn More' });
    
    // Check for hover classes
    expect(image).toHaveClass('transition-transform', 'hover:scale-105');
    expect(link).toHaveClass('hover:bg-blue-700', 'transition-colors');
  });

  // Accessibility test
  it('is accessible', () => {
    render(<SampleProductCard {...defaultProps} />);
    
    // Check for proper heading structure
    const heading = screen.getByRole('heading', { level: 3 });
    expect(heading).toBeInTheDocument();
    expect(heading).toHaveTextContent('Test Product');
    
    // Check for proper link
    const link = screen.getByRole('link');
    expect(link).toBeInTheDocument();
    expect(link).toHaveAccessibleName('Learn More');
    
    // Check for proper image alt text
    const image = screen.getByRole('img');
    expect(image).toHaveAccessibleName('Test Product - Test image description');
  });

  // Different categories test
  it('works with different image categories', () => {
    const categories: ImageCategory[] = ['windows', 'doors', 'siding', 'roofing'];
    
    categories.forEach(category => {
      const { unmount } = render(
        <SampleProductCard {...defaultProps} category={category} />
      );
      
      expect(screen.getByText('Test Product')).toBeInTheDocument();
      unmount();
    });
  });

  // Long description test
  it('handles long descriptions properly', () => {
    const longDescription = 'This is a very long description that should be properly handled by the component. '.repeat(10);
    
    render(
      <SampleProductCard 
        {...defaultProps} 
        description={longDescription}
      />
    );
    
    const description = screen.getByText(longDescription);
    expect(description).toBeInTheDocument();
    expect(description).toHaveClass('line-clamp-3');
  });

  // Empty props test
  it('handles empty title and description gracefully', () => {
    render(
      <SampleProductCard 
        {...defaultProps}
        title=""
        description=""
      />
    );
    
    // Should still render the structure
    expect(screen.getByRole('img')).toBeInTheDocument();
    expect(screen.getByRole('link')).toBeInTheDocument();
  });

  // Special characters test
  it('handles special characters in title and description', () => {
    const specialTitle = 'Test & Product "Special" <Characters>';
    const specialDescription = 'Description with & special "characters" <test>';
    
    render(
      <SampleProductCard 
        {...defaultProps}
        title={specialTitle}
        description={specialDescription}
      />
    );
    
    expect(screen.getByText(specialTitle)).toBeInTheDocument();
    expect(screen.getByText(specialDescription)).toBeInTheDocument();
  });

  // Error boundary test
  it('handles errors gracefully', () => {
    // Mock console.error to avoid noise in test output
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    
    // This should not throw an error even with invalid props
    expect(() => {
      render(<SampleProductCard {...defaultProps} />);
    }).not.toThrow();
    
    consoleSpy.mockRestore();
  });
});
