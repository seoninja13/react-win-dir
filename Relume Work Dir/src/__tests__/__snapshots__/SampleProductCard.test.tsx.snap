// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleProductCard matches snapshot 1`] = `
<div
  class="bg-white rounded-lg overflow-hidden shadow-md hover:shadow-lg transition-shadow "
>
  <div
    class="aspect-video overflow-hidden"
  >
    <img
      alt="Test Product - Test image description"
      class="w-full h-full object-cover transition-transform hover:scale-105"
      src="https://example.com/test-image.jpg"
    />
  </div>
  <div
    class="p-4"
  >
    <h3
      class="text-xl font-semibold mb-2"
    >
      Test Product
    </h3>
    <p
      class="text-gray-600 mb-4 line-clamp-3"
    >
      This is a test product description that should be displayed in the card.
    </p>
    <a
      class="inline-block px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition-colors"
      href="/test-product"
    >
      Learn More
    </a>
  </div>
</div>
`;
